### PR TITLE
Use Git's HTTP proxy configuration

### LINKED ISSUE #11: Support HTTP proxies
The .NET Framework GCM supports being used from behind a proxy. We should look at what is required (if anything) to allow these customer scenarios to work with GCM Core.

### PR DESCRIPTION
**Pull request by mjcheetham**
Use the Git HTTP proxy configuration to create new configured `HttpClient` instances with proxy support.

Introduce the `IWebProxyFactory` component that extracts proxy configuration from Git and the environment to create new `IWebProxy` objects.

Fixes #11

### CODE REVIEW

File: src/shared/Microsoft.Git.CredentialManager/Constants.cs [hunk 1]
```diff
@@ -17,6 +17,10 @@ public static class EnvironmentVariables
             public const string GcmTraceMsAuth     = "GCM_TRACE_MSAUTH";
             public const string GcmDebug           = "GCM_DEBUG";
             public const string GitTerminalPrompts = "GIT_TERMINAL_PROMPT";
+            public const string CurlAllProxy       = "ALL_PROXY";
+            public const string CurlHttpProxy      = "HTTP_PROXY";
+            public const string CurlHttpsProxy     = "HTTPS_PROXY";
+            public const string GcmHttpProxy       = "GCM_HTTP_PROXY";
```
Comment (jeschu1) [inline src/shared/Microsoft.Git.CredentialManager/Constants.cs]: Should Gcm have a specific Proxy Env variable?  Should it come from git instead?

File: src/shared/Microsoft.Git.CredentialManager/HttpClientFactory.cs [hunk 2]
```diff
@@ -20,17 +22,55 @@ public interface IHttpClientFactory
...
+                {
+                    Proxy = proxy,
+                    UseProxy = true
+                };
+            }
+
+            if (!_settings.IsCertificateVerificationEnabled)
+            {
+#if NETFRAMEWORK
+                ServicePointManager.ServerCertificateValidationCallback = (req, cert, chain, errors) => true;
```
Comment (jeschu1) [inline src/shared/Microsoft.Git.CredentialManager/HttpClientFactory.cs]: This is interesting, based on the type of .Net we have to do different settings here?  How fun was this one to find :-)?

File: src/shared/Microsoft.Git.CredentialManager/Settings.cs [hunk 3]
```diff
@@ -55,20 +90,109 @@ public Settings(IEnvironmentVariables environmentVariables, IGit git)
...
+                 *        http.proxy
+                 *
+                 *   3. cURL environment variables
+                 *        HTTPS_PROXY
+                 *        HTTP_PROXY
+                 *        ALL_PROXY
+                 *
+                 *   4. Legacy GCM proxy environment variable
+                 *        GCM_HTTP_PROXY
+                 *
```
Comment (jeschu1) [inline src/shared/Microsoft.Git.CredentialManager/Settings.cs]: Ideally how many ways to set the proxy would we want to support/need?

File: src/shared/Microsoft.Git.CredentialManager/Settings.cs [hunk 4]
```diff
@@ -55,20 +90,109 @@ public Settings(IEnvironmentVariables environmentVariables, IGit git)
...
+
+                bool isHttps = StringComparer.OrdinalIgnoreCase.Equals(Uri.UriSchemeHttps, RemoteUri?.Scheme);
+
+                Uri proxyConfig;
+
+                // 1. GCM proxy Git configuration
+                if (isHttps && TryGetUriSetting(null, GitCredCfg.SectionName, GitCredCfg.HttpsProxy, out proxyConfig) ||
+                    TryGetUriSetting(null, GitCredCfg.SectionName, GitCredCfg.HttpProxy,  out proxyConfig))
+                {
+                    return proxyConfig;
```
Comment (jeschu1) [inline src/shared/Microsoft.Git.CredentialManager/Settings.cs]: Is it worth logging that we used the GitCredCfg.HttpsProxy, setting and what the proxy was?  Same question for other 3 code paths.

File: src/shared/Microsoft.Git.CredentialManager/Settings.cs [hunk 5]
```diff
@@ -55,20 +90,109 @@ public Settings(IEnvironmentVariables environmentVariables, IGit git)
...
+                               TryGetUriSetting(KnownEnvars.CurlAllProxy,   null, null, out proxyConfig))
+                {
+                    return proxyConfig;
+                }
+
+                // 4. Legacy GCM proxy environment variable
+                if (TryGetUriSetting(KnownEnvars.GcmHttpProxy, null, null, out proxyConfig))
+                {
+                    // TODO: warning trace?
+                    // the 'GCM_HTTP_PROXY' variable has been deprecated, use '{ keyName }' instead.
```
Comment (jeschu1) [inline src/shared/Microsoft.Git.CredentialManager/Settings.cs]: I like that idea :-)

File: src/shared/Microsoft.Git.CredentialManager/Trace.cs [hunk 6]
```diff
@@ -337,4 +338,28 @@ private static string FormatText(string message, string filePath, int lineNumber
             return text;
         }
     }
+
+    public static class TraceExtensions
+    {
+        public static void WriteProxyInfo(this ITrace trace, IWebProxy proxy)
+        {
+            var uri = proxy.GetProxy(new Uri("https://example.com"));
```
Comment (jeschu1) [inline src/shared/Microsoft.Git.CredentialManager/Trace.cs]: Can you explain what this function is doing?  Sample output in comments would be helpful.

File: src/shared/Microsoft.Git.CredentialManager/UriExtensions.cs [hunk 7]
```diff
@@ -2,11 +2,37 @@
...
+            EnsureArgument.NotNull(uri, nameof(uri));
+            userName = null;
+            password = null;
+
+            if (string.IsNullOrWhiteSpace(uri.UserInfo))
+            {
+                return false;
+            }
+
+            string[] split = uri.UserInfo.Split(new[] {':'}, count: 2);
```
Comment (jeschu1) [inline src/shared/Microsoft.Git.CredentialManager/UriExtensions.cs]: Can we add a comment with an example of the expected format?

File: src/shared/Microsoft.Git.CredentialManager/HttpClientFactory.cs [hunk 8]
```diff
@@ -20,17 +22,55 @@ public interface IHttpClientFactory
...
+            _settings = settings;
+        }
+
         public HttpClient CreateClient()
         {
-            // Initialize a new HttpClient
-            var client = new HttpClient();
+            HttpClientHandler handler;
+
+            if (TryCreateProxy(out IWebProxy proxy))
```
Comment (jeschu1) [inline src/shared/Microsoft.Git.CredentialManager/HttpClientFactory.cs]: you know what to do here :-)

File: src/shared/Microsoft.Git.CredentialManager/HttpClientFactory.cs (hunk 2 above)
Comment (mjcheetham) [reply src/shared/Microsoft.Git.CredentialManager/HttpClientFactory.cs]: A quick Internet search produced some help on how to configure the proxy settings in both the .NET Core and Framework HTTP stacks.

File: src/shared/Microsoft.Git.CredentialManager/Settings.cs (hunk 3 above)
Comment (mjcheetham) [reply src/shared/Microsoft.Git.CredentialManager/Settings.cs]: Good question.. cc: @jrbriggs

I've gone for maximal compatibility here currently.

File: src/shared/Microsoft.Git.CredentialManager/UriExtensions.cs (hunk 7 above)
Comment (mjcheetham) [reply src/shared/Microsoft.Git.CredentialManager/UriExtensions.cs]: Added.

File: src/shared/Microsoft.Git.CredentialManager/Settings.cs (hunk 5 above)
Comment (mjcheetham) [reply src/shared/Microsoft.Git.CredentialManager/Settings.cs]: Added.

File: src/shared/Microsoft.Git.CredentialManager/Constants.cs (hunk 1 above)
Comment (mjcheetham) [reply src/shared/Microsoft.Git.CredentialManager/Constants.cs]: This is just here for compatibility purposes. We warn about using this deprecated option.

File: docs/netconfig.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+If you have special characters (as defined by [RFC 3986 section 2.2](https://tools.ietf.org/html/rfc3986#section-2.2)) in your username or password such as `:`, `@`, or any other non-URL friendly character you can URL-encode them ([section 2.1](https://tools.ietf.org/html/rfc3986#section-2.2)).
+
+For example, a space character would be encoded with `%20`.
+
+### Other proxy options
+
+GCM Core supports other ways of configuring a proxy for convenience and compatibility.
+
+1. GCM-specific configuration options (_**only** respected by GCM_):
+   - `credential.httpProxy`
```
Comment (jeschu1) [inline docs/netconfig.md]: is it worth mentioning this is being deprecated here too?

File: src/shared/Microsoft.Git.CredentialManager.Tests/HttpClientFactoryTests.cs [hunk 10]
```diff
@@ -22,12 +26,105 @@ public void HttpClientFactory_GetClient_SetsDefaultHeaders()
...
+            {
+                RemoteUri = repoRemoteUri,
+                RepositoryPath = repoPath
+            };
+            var httpFactory = new HttpClientFactory(Mock.Of<ITrace>(), settings, Mock.Of<IStandardStreams>());
+
+            bool result = httpFactory.TryCreateProxy(out IWebProxy proxy);
+
+            Assert.False(result);
+            Assert.Null(proxy);
```
Comment (jeschu1) [inline src/shared/Microsoft.Git.CredentialManager.Tests/HttpClientFactoryTests.cs]: Is it worth verifying any fields on the proxy?

File: src/shared/Microsoft.Git.CredentialManager.Tests/SettingsTests.cs [hunk 11]
```diff
@@ -191,6 +191,266 @@ public void Settings_IsSecretTracingEnabled_EnvarFalsey_ReturnsFalse()
...
+             // Test case 1: cURL environment variables > GCM_HTTP_PROXY
+            envarDict[Constants.EnvironmentVariables.GcmHttpProxy] = value1.ToString();
+            envarDict[Constants.EnvironmentVariables.CurlHttpProxy] = value2.ToString();
+            RunTest(value2);
+
+             // Test case 2: http.proxy > cURL environment variables
+            configDict[$"{Constants.GitConfiguration.Http.SectionName}.{Constants.GitConfiguration.Http.Proxy}"] = value3.ToString();
+            RunTest(value3);
+
+             // Test case 3: credential.httpProxy > http.proxy
```
Comment (jeschu1) [inline src/shared/Microsoft.Git.CredentialManager.Tests/SettingsTests.cs]: did we skip 4 cause it's being depricated?  If so I'm fine with that.

File: src/shared/Microsoft.Git.CredentialManager/CommandContext.cs [hunk 12]
```diff
@@ -112,52 +95,7 @@ public CommandContext()
...
-                    {
-                        AutoFlush = true,
-                        NewLine = LineFeed,
-                    };
-                }
-
-                return _stdErr;
-            }
-        }
+        public IStandardStreams Streams { get; }
```
Comment (jeschu1) [inline src/shared/Microsoft.Git.CredentialManager/CommandContext.cs]: There are many changes for the Streams update.  I really like that you put the changes for this in it's own commit with details.  If you hadn't nicely done that I think would have asked for a separate PR.  But since you did go you :-)!

File: src/shared/Microsoft.Git.CredentialManager/HttpClientFactory.cs [hunk 13]
```diff
@@ -20,17 +23,68 @@ public interface IHttpClientFactory
...
         {
-            // Initialize a new HttpClient
-            var client = new HttpClient();
+            _trace.WriteLine("Creating new HTTP client instance...");
+
+            HttpClientHandler handler;
+
+            if (TryCreateProxy(out IWebProxy proxy))
+            {
+                _trace.WriteLine("HTTP client is using the configured proxy.");
```
Comment (jeschu1) [inline src/shared/Microsoft.Git.CredentialManager/HttpClientFactory.cs]: To verify do we print out the details of the proxy being using anywhere?  That would be helpful for debugging i think.

File: src/shared/Microsoft.Git.CredentialManager/Settings.cs (hunk 4 above)
Comment (mjcheetham) [reply src/shared/Microsoft.Git.CredentialManager/Settings.cs]: We log and warn the user that a legacy setting is being used now (including a link to how to configure the supported ways).

File: docs/netconfig.md (hunk 9 above)
Comment (mjcheetham) [reply docs/netconfig.md]: The line before this says:

> ..for convenience and compatibility.

Perhaps I should make it even more explicit still?

File: src/shared/Microsoft.Git.CredentialManager.Tests/HttpClientFactoryTests.cs (hunk 10 above)
Comment (mjcheetham) [reply src/shared/Microsoft.Git.CredentialManager.Tests/HttpClientFactoryTests.cs]: There are no other fields on `IWebProxy` sadly (without doing some suspicious casting and/or reflection. 😢

File: src/shared/Microsoft.Git.CredentialManager/HttpClientFactory.cs (hunk 13 above)
Comment (mjcheetham) [reply src/shared/Microsoft.Git.CredentialManager/HttpClientFactory.cs]: Yes, we print out the proxy settings used each time a new `HttpClient` instance is created.

Comment (mjcheetham) [reply src/shared/Microsoft.Git.CredentialManager/HttpClientFactory.cs]: https://github.com/mjcheetham/Git-Credential-Manager-Core/blob/28d9cb68dacfb724414ff876b391e68e2d9792c1/src/shared/Microsoft.Git.CredentialManager/HttpClientFactory.cs#L145

**Comment by mjcheetham** [comment]
> My main thing to double check is that we trace the proxy somewhere.

In the `HttpClientFactory` we trace the proxy information when creating and configuring a new `HttpClient` instance.

https://github.com/mjcheetham/Git-Credential-Manager-Core/blob/28d9cb68dacfb724414ff876b391e68e2d9792c1/src/shared/Microsoft.Git.CredentialManager/HttpClientFactory.cs#L145

> Bonus points if we say which path we used to set it.

Sadly we don't know who called us at this level beyond if the method was a deprecated one or not.

File: src/shared/Microsoft.Git.CredentialManager.Tests/SettingsTests.cs (hunk 11 above)
Comment (mjcheetham) [reply src/shared/Microsoft.Git.CredentialManager.Tests/SettingsTests.cs]: We did not skip any. This test is checking the precedence of the 4 options:

```
                     1            2            3
credential.httpProxy > HTTP_PROXY > http.proxy > GCM_HTTP_PROXY
```

There are three inequalities to check for these four options.

File: docs/netconfig.md (hunk 9 above)
Comment (jeschu1) [reply docs/netconfig.md]: I think it doesn't hurt to just have (deprecated) in the notes
